packagedb/management/commands/fix_purl_values.py

```diff
@@ -0,0 +1,226 @@
+class MemorySavingQuerysetIterator(object):
+    def __init__(self,queryset,max_obj_num=1000):
+        self._base_queryset = queryset
+        self._generator = self._setup()
+        self.max_obj_num = max_obj_num
+
+    def _setup(self):
+        for i in range(0,self._base_queryset.count(),self.max_obj_num):
+            # By making a copy of of the queryset and using that to actually access
+            # the objects we ensure that there are only `max_obj_num` objects in
+            # memory at any given time
+            smaller_queryset = copy.deepcopy(self._base_queryset)[i:i+self.max_obj_num]
+            logger.debug('Grabbing next %s objects from DB' % self.max_obj_num)
+            for obj in smaller_queryset.iterator():
+                yield obj
+
+    def __iter__(self):
+        return self._generator
+
+    def next(self):
+        return self._generator.next()
+
+
+def query_sha1_on_maven(sha1, timeout=DEFAULT_TIMEOUT):
+    maven_api_search_url = f'https://search.maven.org/solrsearch/select?q=1:{sha1}'
+    try:
+        response = session.get(maven_api_search_url, timeout=timeout)
+        response.raise_for_status()
+    except (requests.RequestException, ValueError, TypeError) as exception:
+        logger.debug(f"[Exception] {exception}")
+        return False
+    if not response.ok:
+        return f"API query failed for: {maven_api_search_url}"
+    contents = response.json()
+    resp = contents.get('response', {})
+    matched_artifacts = []
+    if resp.get('numFound', 0) > 0:
+        for matched_artifact in resp.get('docs', []):
+            namespace = matched_artifact.get('g', '')
+            name = matched_artifact.get('a', '')
+            version = matched_artifact.get('v', '')
+            ec = matched_artifact.get('ec', [])
+            if not namespace and name and version:
+                continue
+            matched_artifacts.append(
+                MavenArtifact(
+                    namespace=namespace,
+                    name=name,
+                    version=version,
+                    ec=ec,
+                )
+            )
+    return matched_artifacts
+
+
+class Command(VerboseCommand):
+    help = 'Update maven Package download_url values'
+
+    def handle(self, *args, **options):
+        maven_packages = Package.objects.filter(type='maven', sha1__is_null=False)
+        maven_packages_count = maven_packages.count()
+        logger.info(f'Checking {maven_packages_count:,} Maven Package PackageURL values')
+        packages_to_delete = []
+        unsaved_packages = []
+
+        processed_packages_count = 0
+        for i, package in enumerate(MemorySavingQuerysetIterator(maven_packages)):
+            matched_artifacts = query_sha1_on_maven(package.sha1)
+            if not matched_artifacts:
+                # Remove this package from the database because it's not on maven
+                packages_to_delete.append(package)
```

Comment:
Interesting, I assume that since we're not sure where these packages are coming from, we're getting rid of them. Does that mean we should also get rid of those Maven packages where we don't have an SHA1 checksum because we can't be sure where those are coming from either?

Reply:
@keshav-space It may be worth looking up the package by purl values to see if there is a .sha1 or other checksum file for that Maven JAR. I think the intent of the `if not matched_artifacts` line is to get rid of packages with incorrect sha1s.

Reply:
@JonoYang thanks for the clarification. It makes sense, we don't want packages with incorrect SHA1 that would definitely mess up our D2D matches.